crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -1222,6 +1256,68 @@ impl CheckpointBuilder {
         Ok(highest_sequence)
     }
 
+    async fn construct_and_execute_settlement_transactions(
+        &self,
+        sorted_tx_effects_included_in_checkpoint: &[TransactionEffects],
+        checkpoint_height: CheckpointHeight,
+    ) -> (TransactionKey, Vec<TransactionDigest>) {
+        let _scope =
+            monitored_scope("CheckpointBuilder::construct_and_execute_settlement_transactions");
+
+        let tx_key =
+            TransactionKey::AccumulatorSettlement(self.epoch_store.epoch(), checkpoint_height);
+
+        let settlement_txns: Vec<_> = create_accumulator_update_transactions(
+            &self.epoch_store,
+            checkpoint_height,
+            Some(self.effects_store.as_ref()),
+            sorted_tx_effects_included_in_checkpoint,
+        )
+        .into_iter()
+        .map(|tx| {
+            VerifiedExecutableTransaction::new_system(
+                VerifiedTransaction::new_system_transaction(tx),
+                self.epoch_store.epoch(),
+            )
+        })
+        .collect();
+
+        let settlement_digests: Vec<_> = settlement_txns.iter().map(|tx| *tx.digest()).collect();
+
+        debug!(
+            ?settlement_digests,
+            ?tx_key,
+            "created settlement transactions"
+        );
+
+        self.epoch_store
+            .notify_settlement_transactions_ready(tx_key, settlement_txns);
+
+        let settlement_effects = tokio::time::timeout(Duration::from_secs(10), async {
+            self.effects_store
+                .notify_read_executed_effects(&settlement_digests)
+                .await
+        })
+        .await
+        .unwrap_or_else(|_| {
+            fatal!(
```

Comment:
This seems a bit fragile. We could do a debug_fatal but we should be able to return error and retry? It's hard to say for sure that in an extremely overwhelmed scenario this can never be more than 10s.